drivers:platform:stm32:Add API for TDM DMA read.

jansunil opened the pull request:
1) Added no_OS driver API for TDM DMA read
2) Added no_OS driver API for Stop TDM DMA read
3) Added TDM DMA Read in STM32 platform driver
4) Added function for Stop TDM DMA read in STM32 platform driver.

jansunil commented:
Hi Team- Any review comments on this?

buha commented on include/no_os_tdm.h:
```diff
@@ -107,8 +107,12 @@ struct no_os_tdm_platform_ops {
 	/** TDM read operation function pointer */
 	int32_t (*tdm_ops_read)(struct no_os_tdm_desc *, void *, uint16_t);
+	/** TDM DMA read operation function pointer */
+	int32_t (*tdm_ops_read_dma)(struct no_os_tdm_desc *, void *, uint16_t);
```
So these read_dma/write_dma ops are direct mappings of STM32 HAL in this no-OS top level include file. I think we shouldn't do this. In this file we need to be as generic as possible to cover all possible TDM peripherals and what we can do with them.

A better option would be to keep the read/write functions as they are but:
- add ops for stop()
- add ops for pause()
- add ops for resume()
- in no_os_tdm_init_param add callback for void tx_complete_callback(void *arg) and a void *tx_arg
- in no_os_tdm_init_param add callback for void rx_complete_callback(void *arg) and a void *rx_arg

All these new members can be left uninitialized and read()/write() will have normal blocking behavior.
However, if a callback is initialized, read()/write() should have non-blocking behavior.

Pause/Stop/Resume have no meaning in blocking mode so they can either silently return 0 or -EBADSTATE.

jansunil replied:
Do you suggest something similar to the [UART implementation](https://github.com/analogdevicesinc/no-OS/blob/master/drivers/platform/stm32/stm32_uart.c#L130) here?

buha replied:
Exactly !

jansunil replied:
Added ops for stop/pause and resume explicitly.
Added a structure member for the callback functions as well.

buha commented on drivers/platform/stm32/stm32_tdm.c:
```diff
@@ -219,3 +221,54 @@ int32_t stm32_tdm_read(struct no_os_tdm_desc *desc,
+ * @param nb_samples - Number of samples to read.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t stm32_tdm_read_dma(struct no_os_tdm_desc *desc, void *data,
```
...continuation of my other comment:

This way, you will not implement new ops here, but rather modify existing read/write ops.
In them you will be looking of callbacks are defined and behave differently based on that.

```
if (desc->ops->tx_complete_callback)
    HAL_SAI_Receive_DMA();
else
    HAL_SAI_Receive();
```

jansunil replied:
Got your point here.

jansunil replied:
The callback defined check has been made with ``desc->rx_complete_callback``. Would it be relevant to add the callback and make the check with the no_os_tdm_desc?

jansunil commented:
V2: Code changes as per review comments:
1) Removed the stm32_tdm_read_dma() function and added a condition check in the stm32_tdm_read() with the parameter ``rx_complete_callback``

2) Added a ``rx_complete_callback`` and a ``tx_complete_callback`` in the no_os_tdm_init_param.

3) Added the callback parameter in the no_os_tdm_desc so that the parameter check can be made in the stm32_tdm_read() function

jansunil commented:
Hi @buha - are there any discrepancies between the proposed structure and the latest update?

buha commented on include/no_os_tdm.h:
```diff
@@ -109,6 +117,12 @@ struct no_os_tdm_platform_ops {
 	/** TDM write operation function pointer */
 	int32_t (*tdm_ops_write)(struct no_os_tdm_desc *, void *, uint16_t);
+	/** Pause TDM DMA transfer */
+	int32_t (*tdm_ops_pause_dma)(struct no_os_tdm_desc *);
```
I'd drop the `_dma` suffix, it is unnecessary. Technically the implementation of the non-blocking TDM read/write could be done with just interrupts and memcpy's instead of DMA and this generic API needs to cover that case too.

jansunil replied:
Noted. The "_dma" suffix has been removed everywhere

buha commented on drivers/platform/stm32/stm32_tdm.c:
```diff
@@ -219,3 +228,46 @@ int32_t stm32_tdm_read(struct no_os_tdm_desc *desc,
+
+	if (!desc)
+		return -EINVAL;
+
```
I think an extra check is needed here:

```
if (blocking)
    return -ENOSYS;
```

buha replied:
it is up to you how you detect if user configured the API in blocking or non-blocking

buha replied:
(and this comment applies to the implementations of pause and resume too)

jansunil replied:
The blocking check has been made with respect to the definition of rx_callback.

buha commented on drivers/platform/stm32/stm32_tdm.c:
```diff
@@ -219,3 +228,46 @@ int32_t stm32_tdm_read(struct no_os_tdm_desc *desc,
+ */
+int32_t stm32_pause_tdm_dma_transfer(struct no_os_tdm_desc *desc)
+{
+	return 0;
```
resume/pause need to be implemented just like the stop

jansunil replied:
Resume and Pause are also implemented now.

buha reviewed (commented):
looks good, we're almost there in my opinion

buha commented on drivers/platform/stm32/stm32_tdm.c:
```diff
@@ -129,6 +132,10 @@ int32_t stm32_tdm_init(struct no_os_tdm_desc **desc,
 		goto error;
 		break;
 	};
+	if (param->rx_complete_callback) {
```
you don't need to check, you can simply copy it, if it's NUL you will anyway detect it later when checking `desc->rx_complete_callback`

jansunil replied:
Removed the check here

buha commented:
One other thing is that the tx and rx callbacks are not used. Also, their argument should be passed by the user in init_param. The callbacks are provided by the user and they should be called by your implementation (in stm32_tdm.c) upon transfer completion.

But I'm still cooking the details of the implementation, I need to consult with the team.

buha commented:
So @jansunil ,
I think this is how it should be implemented.
As you already know, UART implements something similar, we can use that as a template.

Initialize the DMA Tx Complete interrupts as it is done here:
https://github.com/analogdevicesinc/no-OS/blob/master/drivers/platform/stm32/stm32_uart.c#L135-L156

Of course, for this to work you need to insert two new entries here, NO_OS_EVT_TDM_TX_COMPLETE, NO_OS_EVT_TDM_RX_COMPLETE and make the correlation with HAL_SAI_TX_COMPLETE_CB_ID, HAL_SAI_RX_COMPLETE_CB_ID
https://github.com/analogdevicesinc/no-OS/blob/8708a859e2c7a8d69c57db0905f8638d83132aa9/drivers/platform/stm32/stm32_irq.c#L68

And of course, a new entry is required here, for the TDM:
https://github.com/analogdevicesinc/no-OS/blob/7fc98116d819e9c0b0bec4cff8986f96f553d79e/include/no_os_irq.h#L87

buha commented:
All this, so that the user-passed tx/rx complete callbacks will be called when the events occur.

jansunil commented:
Thanks for this @buha . Let me work on this and get back.

buha commented:
But really, you should test this on your side, I have no setup that could test this, so make sure the interrupts work and that the callbacks get called ;)

buha commented:
@jansunil any updates on this ?

jansunil commented:
Hi @buha - May I take some more time to get back on this?

buha commented:
sure, of course

jansunil commented:
V3: Implemented SAI DMA callbacks
Added register callback for SAI DMA similar to UART interrupt mode.

jansunil commented:
V4: Fixed build errors

jansunil commented:
Hi @buha ,

Apologies for getting back late on this activity. The PR is ready for review now.

buha reviewed (approved):
Looks good, I think this will work.

I still have a question about what happens if we use DMA in a project for both (for example) SPI, UART and SAI - how will we differentiate between them ?

buha commented:
I'm merging this anyway, we can deal with that separately.
